### ISSUE TITLE
Python 3.10 macOS universal2 wheel is not universal (does not contain arm64 architecture)

### ISSUE THREAD

**Issue by burnpanck** [Unidata/netcdf4-python#1148]
Installing on a M1 mac with python 3.10 using pip succeeds.
```
$ pip install netcdf4
Collecting netcdf4
  Downloading netCDF4-1.5.8-cp310-cp310-macosx_10_9_universal2.whl (4.2 MB)
     |████████████████████████████████| 4.2 MB 4.8 MB/s            
Requirement already satisfied: numpy>=1.9 in ./.pyenv/versions/3.10.0/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from netcdf4) (1.21.4)
Requirement already satisfied: cftime in ./.pyenv/versions/3.10.0/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from netcdf4) (1.5.1.1)
Installing collected packages: netcdf4
Successfully installed netcdf4-1.5.8
```
However, the resulting install is broken:
```
$ python -c "import sys;print(sys.version);import netCDF4"
3.10.0 (default, Dec 17 2021, 20:54:45) [Clang 13.0.0 (clang-1300.0.29.30)]
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/yves/.pyenv/versions/3.10.0/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/netCDF4/__init__.py", line 3, in <module>
    from ._netCDF4 import *
ImportError: dlopen(/Users/yves/.pyenv/versions/3.10.0/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/netCDF4/_netCDF4.cpython-310-darwin.so, 0x0002): tried: '/Users/yves/.pyenv/versions/3.10.0/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/netCDF4/_netCDF4.cpython-310-darwin.so' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64e')), '/usr/local/lib/_netCDF4.cpython-310-darwin.so' (no such file), '/usr/lib/_netCDF4.cpython-310-darwin.so' (no such file)
```

**Comment by burnpanck** [comment]
The same issue also exists for the `cftime` dependency: https://github.com/Unidata/cftime/issues/264

Building from source using dependencies from homebrew works.

**Comment by weatherfrog** [comment]
> Building from source using dependencies from homebrew works.

@burnpanck How does this work? I brew installed `hdf5` and `netcdf` and still get:
```
$ pip install netcdf4==1.5.8
[...]
  Complete output (29 lines):
  Package hdf5 was not found in the pkg-config search path.
  Perhaps you should add the directory containing `hdf5.pc'
  to the PKG_CONFIG_PATH environment variable
  No package 'hdf5' found
  Traceback (most recent call last):
    File "<string>", line 1, in <module>
    File "/private/var/folders/51/gxcnz34j0vs83p9g7zdncqhm0000gn/T/pip-install-2ckr_df_/netcdf4_10dd53610fbd4cca90571921914e1aaa/setup.py", line 409, in <module>
      _populate_hdf5_info(dirstosearch, inc_dirs, libs, lib_dirs)
    File "/private/var/folders/51/gxcnz34j0vs83p9g7zdncqhm0000gn/T/pip-install-2ckr_df_/netcdf4_10dd53610fbd4cca90571921914e1aaa/setup.py", line 360, in _populate_hdf5_info
      raise ValueError('did not find HDF5 headers')
  ValueError: did not find HDF5 headers
  reading from setup.cfg...
  using nc-config...
  checking /opt/homebrew/Cellar/netcdf/4.8.1/include ...
  hdf5 headers not found in /opt/homebrew/Cellar/netcdf/4.8.1/include
  nc-config did provide path to HDF5 headers, search standard locations...
      HDF5_DIR environment variable not set, checking some standard locations ..
  checking /Users/rg/include ...
  hdf5 headers not found in /Users/rg/include
  checking /usr/local/include ...
  hdf5 headers not found in /usr/local/include
  checking /sw/include ...
  hdf5 headers not found in /sw/include
  checking /opt/include ...
  hdf5 headers not found in /opt/include
  checking /opt/local/include ...
  hdf5 headers not found in /opt/local/include
  checking /usr/include ...
  hdf5 headers not found in /usr/include
```

The hdf5 headers are in `/opt/homebrew/Cellar/hdf5/1.12.1/include/`, but the setup routine does not look there.

**Comment by jswhit** [comment]
github actions apparently does not yet provide support for Apple M1 hosted workflow (https://github.com/actions/virtual-environments/issues/2187), and our current method for building binary wheels (https://github.com/MacPython/netcdf4-python-wheels) uses github actions.  Any suggestions on how to deal with this limitation are welcome.

**Comment by dopplershift** [comment]
@jswhit I think the multibuild in that repo needs to be bumped to github.com/multibuild/multibuild (See [this issue](https://github.com/multi-build/multibuild/issues/442)).

**Comment by dopplershift** [comment]
To be clear, I think the build should work fine because Apple's build tools work fine to cross-compile for M1 from an x86_64 machine.

**Comment by weatherfrog** [comment]
I'll answer my own question. Installation does work with
```
HDF5_DIR=/opt/homebrew/Cellar/hdf5/1.12.1 pip install netcdf4==1.5.8
```
Sorry to have kinda hijacked this issue.

**Comment by jswhit** [comment]
thanks @dopplershift - trying that now

**Comment by jswhit** [comment]
@burnpanck could you try a 3.10 wheel from https://github.com/MacPython/netcdf4-python-wheels/releases/tag/v1.5.8rel?

**Comment by burnpanck** [comment]
@jswhit: I don't see any arm or universal wheels there - maybe the github actions need some re-configuring? From a quick glance at [`travis_osx_steps.sh`](https://github.com/multi-build/multibuild/blob/3903f7f29d313d4742483e6af5afd460664afe2c/travis_osx_steps.sh), it seems to expect `$PLAT= universal2` to build those universal wheels (at least there is some conditional code based on this). I don't see any mention of  `$PLAT` related to macos within [the github action script](https://github.com/MacPython/netcdf4-python-wheels/blob/master/.github/workflows/build-wheels.yml). I have to say though that I don't known neither GitHub Actions nor multibuild at all.

**Comment by jswhit** [comment]
@burnpanck  the universal2 wheel is there now

**Comment by recmo** [comment]
Installation no works and seems to use wheel, but it does not work.

```
$ pip install netcdf4 
DEPRECATION: Configuring installation scheme with distutils config files is deprecated and will no longer work in the near future. If you are using a Homebrew or Linuxbrew Python, please see discussion at https://github.com/Homebrew/homebrew-core/issues/76621
Collecting netcdf4
  Using cached netCDF4-1.5.8-cp39-cp39-macosx_12_0_arm64.whl
Requirement already satisfied: cftime in /opt/homebrew/lib/python3.9/site-packages (from netcdf4) (1.5.1.1)
Requirement already satisfied: numpy>=1.9 in /opt/homebrew/lib/python3.9/site-packages (from netcdf4) (1.22.1)
Installing collected packages: netcdf4
  DEPRECATION: Configuring installation scheme with distutils config files is deprecated and will no longer work in the near future. If you are using a Homebrew or Linuxbrew Python, please see discussion at https://github.com/Homebrew/homebrew-core/issues/76621
DEPRECATION: Configuring installation scheme with distutils config files is deprecated and will no longer work in the near future. If you are using a Homebrew or Linuxbrew Python, please see discussion at https://github.com/Homebrew/homebrew-core/issues/76621
Successfully installed netcdf4-1.5.8
```

```
$ python -c "import sys;print(sys.version); import netCDF4"    

3.9.10 (main, Jan 15 2022, 11:40:53) 
[Clang 13.0.0 (clang-1300.0.29.3)]
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/opt/homebrew/lib/python3.9/site-packages/netCDF4/__init__.py", line 3, in <module>
    from ._netCDF4 import *
ImportError: dlopen(/opt/homebrew/lib/python3.9/site-packages/netCDF4/_netCDF4.cpython-39-darwin.so, 0x0002): symbol not found in flat namespace '_H5get_libversion'
```

**Comment by dopplershift** [comment]
@recmo I don't know where this came from:
>  Using cached netCDF4-1.5.8-cp39-cp39-macosx_12_0_arm64.whl
But that is NOT listed at https://github.com/MacPython/netcdf4-python-wheels/releases/tag/v1.5.8rel

(Did you try building locally?) Can you retry with `pip install --no-cache-dir netcdf4` ?

**Comment by jswhit** [comment]
@recmo did you download the universal2 wheel for python 3.10 from  https://github.com/MacPython/netcdf4-python-wheels/releases/tag/v1.5.8rel?  ?  I have not updated the pypi wheels yet.

**Comment by jswhit** [comment]
I have generated updated universal2 wheels for cftime also (at https://github.com/MacPython/cftime-wheels/releases/tag/v1.5.1.1rel).  If someone can confirm these, and the netcdf4-python wheels at  https://github.com/MacPython/netcdf4-python-wheels/releases/tag/v1.5.8rel actually work on Apple M1, I will create new pypi releases for both packages.

**Comment by burnpanck** [comment]
I see the same behaviour @recmo observed:
```
$ pip install ./netCDF4-1.5.8-cp310-cp310-macosx_10_9_universal2.whl
Processing ./netCDF4-1.5.8-cp310-cp310-macosx_10_9_universal2.whl
Requirement already satisfied: numpy>=1.9 in /Users/yves/.pyenv/versions/3.10.0/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from netCDF4==1.5.8) (1.21.4)
Requirement already satisfied: cftime in /Users/yves/.pyenv/versions/3.10.0/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from netCDF4==1.5.8) (1.5.1.1)
Installing collected packages: netCDF4
  Attempting uninstall: netCDF4
    Found existing installation: netCDF4 1.6.0
    Uninstalling netCDF4-1.6.0:
      Successfully uninstalled netCDF4-1.6.0
Successfully installed netCDF4-1.5.8

$ python -c "import sys;print(sys.version);import netCDF4"
3.10.0 (default, Dec 17 2021, 20:54:45) [Clang 13.0.0 (clang-1300.0.29.30)]
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/yves/.pyenv/versions/3.10.0/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/netCDF4/__init__.py", line 3, in <module>
    from ._netCDF4 import *
ImportError: dlopen(/Users/yves/.pyenv/versions/3.10.0/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/netCDF4/_netCDF4.cpython-310-darwin.so, 0x0002): symbol not found in flat namespace '_H5get_libversion'
```

I observe the same behaviour with the `arm64` wheel.

The corresponding `cftime` wheels (both `universal2` and `arm64`) do work as expected though.

**Comment by jswhit** [comment]
Looks like the C libs that are embedded within the wheel are not build with the right cross-compiling flags (only the python c extension is).  Pinging @matthew-brett to see if he has any suggestions on how to get functioning universal2 wheels (including dependent C libs) using multibuild.

**Comment by jswhit** [comment]
Note that as a workaround there does appear to be a working conda netcdf4-python native osx-arm64 package available (https://anaconda.org/conda-forge/netcdf4)

**Comment by jswhit** [comment]
Looks like we won't have a solution for Apple M1 wheels until Github Actions supports that platform natively.  It's just going to be too difficult to cross-compile all the C lib dependencies on x86_64.  Another solution would be to use another CI provide (azure?) that supports this.

**Comment by dopplershift** [comment]
@jswhit I'm not sure what the problem is exactly with cross-compiling. Conda-forge's osx-arm64 packages are made by cross-compiling, not with native M1 builds.

**Comment by jswhit** [comment]
@dopplershift I think the problem is my ignorance - I can't see how to tell multibuild to cross-compile the C library dependencies and bundle them with the arm64 and universal2 macosx wheels.

**Comment by dopplershift** [comment]
@jswhit Yeah, I sense the issue is multibuild. cibuildwheel is another option that's out there (that *think* numpy/scipy is using). I have not dealt with it, so no personal recommendation or expertise to offer.

**Comment by recmo** [comment]
@burnpanck Thank you, that work-around worked for me!

**Comment by jswhit** [comment]
cftime 1.5.2 now has arm64 and universal2 wheels, so I think @burnpanck's workaround can now be simplified to:

```
brew install hdf5 netcdf
git clone https://github.com/Unidata/netcdf4-python.git
HDF5_DIR=/opt/homebrew/Cellar/hdf5/1.12.1 pip install ./netcdf4-python
```

**Comment by jswhit** [comment]
This is now fixed, so the next release (1.6.0) will have working arm64 and universal2 wheels for macosx.    Working wheels for 1.5.8 are available at https://github.com/MacPython/netcdf4-python-wheels/releases/tag/v1.5.8rel.
